fix/fct_weekly_school_acquisition_metrics

bakerfranke opened the pull request:
## Description

This is a major PR that represents the end (hopefully) validating school acquisition (aka school status) in Hydrone v. Rosetta.  The changes committed here, and tests below should represent this as launch-ready.

## Important look-fors

**1. left-padding school_ids with 0s**
This was introduced in a previous iteration of this PR.

**DRY solution** Introduced a macro [dbt/macros/pad_school_id.sql](https://github.com/code-dot-org/analytics/pull/62/files#diff-37eddbde5657c2280652020f4fad46725d742197b59d30db13ddb501d93da1dc) to use to left-pad school_ids so they always have 12 or 8 characters.   The macro was then applied in 3 files close to the source where school_id was coming in.

**2. section mapping**
[dbt/models/intermediate/int_section_mapping.sql](https://github.com/code-dot-org/analytics/pull/62/files#diff-ec9f438ae56e91ec104fb41aef29c95432a9346de128988836daea5fbff2d54d) added code here given by @allison-code-dot-org which I think she gave in a separate PR.  This should not conflict with that.

**3. Student courses**
[dbt/models/marts/students/dim_student_courses.sql](https://github.com/code-dot-org/analytics/pull/62/files#diff-8687b955c015617c1d96c0738fc3922c46724078487d75fa7b7c6b01fe13478a) -- modified so that students only "count" activity toward student-facing courses.  prior iteration was overly-inclusive of all courses.  This change was made to **align with the definition in Rosetta** for now (for launch and validation).

**3. Followers & Sections**
[dbt/models/staging/dashboard/base/base_dashboard__followers.sql](https://github.com/code-dot-org/analytics/pull/62/files#diff-b68b56462d27675840d40ff3ccea7b3e96b83a9959aa98954d78b442c7b625bb) 
dbt/models/staging/dashboard/base/base_dashboard__sections.sql

We were filtering out deleted follower records and sections which was causing us to miss historical activity records.  If we want to filter out deleted follower records it should be done further downstream with an analytical purpose/choice behind it (for example: an analysis of characteristics of deleted followers, or something.)

**4. School Stats by Years**
[dbt/models/staging/dashboard/stg_dashboard__school_stats_by_years.sql](https://github.com/code-dot-org/analytics/pull/62/files#diff-f9adbf799f65b859a44032bb8b1be6f8743f61b8d1634a699a9fcb16c4e3f50d) This is a *major* change to align the results of this model with Rosetta school stats.  The major issue was that it was not classifying schools as el/ml/hi correctly, or even close to what `analysis.school_stats` was showing.  The two major issues were (1) upper/lower case mismatch on value comparisons (2) some faulty logic related to what to check, and in what order, for classifying as stage_el, stage_mi, stage_hi.  

**5. dim_school_status**
File was changed only to cleanup some formatting.  no significant changes in this PR.

**6. active status label**
[dbt/macros/active_status_label.sql](https://github.com/code-dot-org/analytics/pull/62/files#diff-31e1727feea720d78655e6403576d1abfa9b8ebd72688803caff12b1acb0cc7a)

This is an UNIMPLEMENTED IDEA.  Made this macro to DRY program our application of text labels to status codes that we've adopted in several places in our code.  I didn't use use this macro anywhere yet.  Just committing it for potential future use.

## Testing

**metrics match**
This small snippet represents the fact that all of the models and tables beneath this metrics table are correct.

```
select *
from {{ ref('fct_weekly_school_acquisition_metrics') }}
where 
  school_year = '2022-23'
  and school_year_week = 53
order by school_level, status
```
The outputs there ^^ match the final output of the weekly school acquisition metrics we've been reporting in MIR.

**dim_schools v. school_stats**

Compare the counts of schools by school-level between dim_schools and analysis.school_stats.  The work in stg_dashboard__school_stats_by_years achieves this match.sql

Hydrone dim_schools:
```
select 
count(distinct(school_id)), school_level_simple, is_stage_el, is_stage_mi, is_stage_hi
from dev.dbt_baker.dim_schools
group by 2,3,4,5
order by 2
```

Corollary with school_stats:

```
  SELECT
    count(distinct(ss.school_id)),
    (
        (case when ss.stage_el = 1 then 'el' else '__' end ) ||
        (case when ss.stage_mi = 1 then 'mi' else '__' end ) ||
        (case when ss.stage_hi = 1 then 'hi' else '__' end ) 
    ) as ros_school_level_simple,
    stage_el,
    stage_mi,
    stage_hi
  FROM school_stats ss
  GROUP BY 2,3,4,5
  order by 2
  ```

**all school_ids have 8 or 12 characters**
```
SELECT count(distinct(school_id)), len(school_id) from {{ref ('dim_schools')}} group by 2
SELECT count(distinct(school_id)), len(school_id) from {{ref ('stg_dashboard__school_stats_by_years group') }} by 2
SELECT count(distinct(school_id)), len(school_id) from {{ref ('stg_dashboard__school_infos') }} group by 2
SELECT count(distinct(school_id)), len(school_id) from {{ref ('stg_dashboard__schools') }} group by 2
```





## Future work

1. [dbt/macros/pad_school_id.sql](https://github.com/code-dot-org/analytics/pull/62/files#diff-37eddbde5657c2280652020f4fad46725d742197b59d30db13ddb501d93da1dc) could be made a little more robust with some testing added. We should probably put some kind of test in to ensure that all NCES ids being used in the system are 8 or 12 characters.  I'm a little nervous that we may unwittingly onboard data from source in the future where we'll forget to apply this macro.  It should fail to build noisily if so.

2. DRY program the status labels (see macro active_status_label.sql)

3. Should fast follow with tickets to **add documentation** around these tables.  Will do as part of launch.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code adheres to [style guide](https://docs.getdbt.com/best-practices/how-we-style/0-how-we-style-our-dbt-projects)👀 and is **[DRY](https://docs.getdbt.com/terms/dry)**
- [ ] Code is well-commented (please refer to style guide for commenting conventions^)
- [ ] New features are translatable or updates will not break up/downstream models
- [ ] Relevant documentation has been added or updated (i.e. `dbt docs` has been published to [site](code-dot-org.github.io/analytics/).
- [ ] Pull Request is labeled appropriately (eg. `chore/`, `feature/`, `fix/`)
- [ ] Follow-up work items (including potential tech debt) are tracked and linked (if applicable)

allison-code-dot-org reviewed (commented):
This seems good as a solution- I'm only wondering why we chose to not do this before. I can't remember the exact reasons, but I remember @nataliazm99 and @jordan-springer and I deciding against it? Maybe Natalia will have more context here.

jordan-springer reviewed (changes requested):
Looks good, few changes and follow-up

bakerfranke commented on dbt/macros/pad_school_id.sql:
```diff
@@ -0,0 +1,21 @@
+-- Left pad school_id with 0s
+{% macro pad_school_id(school_id) %}
+    case
+        when length({{ school_id }}) = 11 then lpad({{ school_id }}, 12, '0')
```
school_id length can either be 8 or 12.  Agree we should add a test.  the real risk comes in when onboarding NCES data each year.  If we can correct it coming in the door - that's where the test should be.

jordan-springer replied:
^ I agree, and we should probably address this in the /base model

bakerfranke commented on dbt/models/staging/dashboard/stg_dashboard__school_infos.sql:
```diff
@@ -3,6 +3,25 @@ school_infos as (
-select * 
-from school_infos
\ No newline at end of file
+        school_info_id,
```
tl;dr - school_info_id is an id our application gives to a school-as-entered-by-a-teacher.  If that school matches/can be found in the NCES set of schools a `school_id` is associated with it as well.  Not all schools are US/NCES schools so this is our system for capturing them.

bakerfranke commented on dbt/models/staging/dashboard/base/base_dashboard__sections.sql:
```diff
@@ -32,7 +32,8 @@ renamed as (
         updated_at,
-        first_activity_at
+        first_activity_at,
+        deleted_at
```
noted. done.

bakerfranke commented:
@jordan-springer @allison-code-dot-org 

I updated and renamed this PR to match the branch I was working on (I had sent PR for just the school_id padding macro.  But now this is the major PR to complete the validation of school acquisition metrics).

I've significantly updated the front page/description.  I'm a bit at a loss for tests.  I put a put a few testing cases in the doc but it might be trivial.  

But I think this is ready for review.

**NOTE**
My preference is that if there is non-functional work (like fixing/adding documentation, styling nits, etc) that it be turned into jira tickets for fast-follow cleanup work.  **We need this merged so that we can continue making correct data/dashboards collateral for the launch**

jordan-springer commented on dbt/models/staging/dashboard/stg_dashboard__school_stats_by_years.sql:
```diff
@@ -91,65 +92,65 @@ school_stats_by_years_adjusted as (
-                is_grade_04,
-                is_grade_05
-                ) = 1 then 1
+        case 
```
@bakerfranke are both of these necessary? why did we remove the coalesce statement?

jordan-springer reviewed (commented):
looks good, a few things to update still but I will add in a separate commit

bakerfranke commented on dbt/models/staging/dashboard/stg_dashboard__school_stats_by_years.sql:
```diff
@@ -91,65 +92,65 @@ school_stats_by_years_adjusted as (
-                is_grade_04,
-                is_grade_05
-                ) = 1 then 1
+        case 
```
Yes, both are necessary.  The data we get from NCES is...messy.  In theory it gives both the low and highest grade served, as well as 0/1 for each grade level.  However, sometimes it doesn't, sometimes it's null, sometimes it's incomplete...and a mix of all those.  

Largely what I'm trying to do here is make hydrone schools match rosetta school stats - which it does.  I used the logic in the existing school stats to identify cases where the prev version of hydrone was missing or misallocating schools.

bakerfranke replied:
The coalesce fails in cases where elementary schools serve some elementary grades but not all because it reports 0/1 as well as null/1 for these fields.  For example if some school serves only grades 1-5, then pk=0 and kg=0.   So all grade levels are: 0,0,1,1,1,1,1

We want to call that an elementary school but if you coalese those values 0 is the first non-null value so it reports it as '0' -- not an elementary school.   Coalese would work if it was: null, null, 1,1,1,1,1, but not all cases are like that. So i replaced it with a big OR.

bakerfranke commented on dbt/models/staging/dashboard/stg_dashboard__school_stats_by_years.sql:
```diff
@@ -91,65 +92,65 @@ school_stats_by_years_adjusted as (
+                OR is_grade_04 = 1 
+                OR is_grade_05 = 1 
+            then 1
+            when grades_offered_lo is null then null -- I don't this is possible unless everything else is already null
```
It is probably safe to remove this line.  I kept it in because analysis.school_stats has it...and I was trying to make sure it was consistent.  Everything matches now so i was relucted to remove and re-validate.

bakerfranke reviewed (commented):
@jordan-springer - I responded to your new questions.
